src/ucp/api/ucp.h

```diff
@@ -369,19 +369,20 @@ enum ucp_dt_type {
  * ucp_mem_map() function.
  */
 enum {
-    UCP_MEM_MAP_NONBLOCK = UCS_BIT(0), /**< Complete the mapping faster, possibly by
-                                            not populating the pages in the mapping
-                                            up-front, and mapping them later when
-                                            they are accessed by communication
-                                            routines. */
-    UCP_MEM_MAP_ALLOCATE = UCS_BIT(1), /**< Identify requirement for allocation,
-                                            if passed address is not a null-pointer
-                                            then it will be used as a hint or direct
-                                            address for allocation. */
-    UCP_MEM_MAP_FIXED    = UCS_BIT(2)  /**< Don't interpret address as a hint:
-                                            place the mapping at exactly that
-                                            address. The address must be a multiple
-                                            of the page size. */
+    UCP_MEM_MAP_NONBLOCK   = UCS_BIT(0), /**< Complete the mapping faster, possibly by
+                                              not populating the pages in the mapping
+                                              up-front, and mapping them later when
+                                              they are accessed by communication
+                                              routines. */
+    UCP_MEM_MAP_ALLOCATE   = UCS_BIT(1), /**< Identify requirement for allocation,
+                                              if passed address is not a null-pointer
+                                              then it will be used as a hint or direct
+                                              address for allocation. */
+    UCP_MEM_MAP_FIXED      = UCS_BIT(2), /**< Don't interpret address as a hint:
+                                              place the mapping at exactly that
+                                              address. The address must be a multiple
+                                              of the page size. */
+    UCP_MEM_MAP_DEVICE_MEM = UCS_BIT(3)  /**< Allocate memory on device */
```

Comment:
Since this is new type of memory we have to be more specific about permissions, coherency definition, and memory ordering semantics. From the the above it is not clear what is this type of memory. We have to define this.

Reply:
added some explanation of device memory

Reply:
@yosefe do you want directly expose non cacheable memory to user space ? Obviously memcpy and other usespace are not optimized to work with this type of memory. Also reads might be expensive. The update description is not good enough and very x86 specific. We would have to make it generic and architecture agnostic.

Reply:
Does even makes sense to expose that address without exposing specialized pack/unpack function ?

Reply:
@shamisp i think the latest documentation is pretty extensive and architecture-agnostic. can you pls suggest alternative documentation which you think would be better?

Reply:
@yosefe it is nowhere agnostic. Moreover we pretty much expose new memory model here. We need really solid justification why we are doing this and the model how user are going to use the memory. I don't think it is useful without pack and unpack functions.

Reply:
@shamisp which part is not agnostic? AFAIK write-combining is a concept which is relevant for many CPU architectures, and even used by this name in internal Linux kernel APIs.
Also this is not the first library to expose such memory to userspace: see gdrcopy for example.
For now, we can mention this memory can't be accessed directly, but only used as target for RMA/AMO operations.

Reply:
> @yosefe it is nowhere agnostic. Moreover we pretty much expose a new memory model here. We need really a solid justification why we are doing this and the model how user are going to use the memory. I don't think it is useful without pack and unpack functions.

@shamisp Probably, I'm missing something. We already expose and use more than one memory model and type -- Host memory, GPU's device memory, strict memory model (x86), relaxed consistency model (POWER, Arm). As long as we can ensure correct and consistent semantics to the User, I don't see how this is different. The interfaces ensures the semantics
and it can rely on varying implementations.

Reply:
@yosefe 

1. In terms of wording I would use the following definitions:
```
 Allocated buffer is
mapped as Write combining buffer which
cannot be used for general memory access
(data or code regions) due to the weak
ordering. Write-combining does not
 guarantee that the combination of writes
and reads is done in the expected order.
```
->

``` 
Allocated buffer is typically mapped as non-cacheable memory region
with memory store gathering capabilities. The memory semantics (ordering, alignment 
requirements, speculative access) of the memory is different from normal cacheable memory 
semantic ("host memory") defined by underlaying architecture. User cannot 
access the memory buffer directly and must use UCP defined interfaces:
 <reference to interfaces>
```

2. We need to define a new UCP interface to access the memory. You can use ```gdrcopy``` API as a reference: https://github.com/NVIDIA/gdrcopy/blob/master/gdrapi.h#L116

3. I want to see a use case (pseudo code) how the API is expected to be used.

Reply:
```
Allocate memory on device. It allows
using on-chip memory, located on the
device, as a data buffer for send/receive
and RDMA operations. Using the device
memory to store packets for transmission
can reduce transmission latency compared
to the host memory. 
```
->
```Allocate memory on device. It allows
using  the device memory as a data buffer for UCP communication routines. 
Using the device memory can reduce transmission latency compared to 
the host memory.
````
@tonycurtis probably can word it better :-).

Reply:
@shamisp On documentation, it says more than required. In fact, we should not say anything about performance and consistency model. We should only caution to make no assumptions (in fact, this is true for all other allocations, however, that is a discussion for different day)

I think we could replace all documentation with this. 

`The memory is allocated on the device. No assumptions should be made about the consistency and the coherency model of the memory. However, allocating the buffers on the device memory and using through the UCP interfaces, the interfaces’ semantics are honored.`

Reply:
@manjugv I just highlighted potentially differences in the memory semantics. The goal is to emphasize why there is special allocator and memory movement functions for this. Probably we can improve the language (through the entire document :-) )

Reply:
@shamisp I disagree. What any good interface documentation should say is the guarantees provided between the Interface User and the Interface Implementation. If you want to tell all details, it should belong to a different document. Also, in this case memory semantics change between architectures and over time the device memory implementation might change. What does not change is the guarantees provided.

Reply:
@manjugv go ahead and make a proposal for the text. It should go to experimental API. I don't think they API is acceptable without definition of API for accessing the memory. If user space does not access the memory we should not expose it all together.

Reply:
Already did above : https://github.com/openucx/ucx/pull/3250#discussion_r261616392